### PR TITLE
Codechange: Use std::vector for midifile's ByteBuffer.

### PR DESCRIPTION
**Pull request by PeterN**
## Motivation / Problem

MidiFile's ByteBuffer uses malloc/free behind the scenes, along with a member to record the length.

## Description

Replace the backing store with a std::vector<byte>.

## Limitations

The full MidiFile parser seems to only used with the Windows MIDI drivers, so while this is compiled on other OSes, functionally it is not tested.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/music/midifile.cpp [hunk 1]
```diff
@@ -149,8 +139,8 @@ class ByteBuffer {
 	bool ReadBuffer(byte *dest, size_t length)
 	{
 		if (this->IsEnd()) return false;
-		if (this->buflen - this->pos < length) return false;
-		memcpy(dest, this->buf + this->pos, length);
+		if (this->buf.size() - this->pos < length) return false;
+		std::copy(std::begin(this->buf) + this->pos, std::begin(this->buf) + this->pos + length, dest);
```
Comment (glx22) [inline src/music/midifile.cpp:143]: Can't use `this->buf.begin()` ?

File: src/music/midifile.cpp [hunk 2]
```diff
@@ -164,8 +154,8 @@ class ByteBuffer {
 	bool ReadDataBlock(MidiFile::DataBlock *dest, size_t length)
 	{
 		if (this->IsEnd()) return false;
-		if (this->buflen - this->pos < length) return false;
-		dest->data.insert(dest->data.end(), this->buf + this->pos, this->buf + this->pos + length);
+		if (this->buf.size() - this->pos < length) return false;
+		dest->data.insert(dest->data.end(), std::begin(this->buf) + this->pos, std::begin(this->buf) + this->pos + length);
```
Comment (glx22) [inline src/music/midifile.cpp:158]: Same here

**Review by glx22 — COMMENTED** [review]
Tested locally, I can hear music.

File: src/music/midifile.cpp (hunk 1 above)
Comment (PeterN) [reply src/music/midifile.cpp:143]: My understanding is that std::begin() is preferred in modern C++.

Comment (glx22) [reply src/music/midifile.cpp:143]: I see, it's more generic, and allows to easily change `this->buf` underlying type.

**Review by glx22 — APPROVED** [review]

**Pull request merged**